### PR TITLE
Faster running median

### PR DESCRIPTION
**Pull request by finagle29**
This PR uses `bottleneck.move_median` to speed up fast_running_median by a factor of ~16 in a simple test I conducted on random data. In a simple real world test, this sped up calibrations by 30% in a quicklook-type DBSPr setup with 1 arc and 1 flat frame.

Additionally, while testing to see if this behaves the same as scipy.ndimage.median_filter and the old fast_running_median, I discovered a bug in scipy.ndimage.median_filter and the previous fast_running_median: when using an even window size, the average of the two middle elements is not taken, instead one of the middle elements is picked.

Example:
```python
>>> seq = np.array([2,1]*2)
>>> # old fast_running_median
>>> ret1 = fast_running_median(seq, 2)
>>> # new fast_running_median (bn = bottleneck)
>>> ret2 = fast_running_median_bn(seq, 2)

>>> print(seq)
[2 1 2 1]
>>> print(ret1)
[2 2 2 2]
>>> print(ret2)
[2.  1.5 1.5 1.5]
```
image

The new fast_running_median_filter works as expected here, as well as matching the old behavior exactly for odd window size

```python
>>> seq = np.random.random(1000)
>>> ret1 = fast_running_median(seq, 413)
>>> ret2 = fast_running_median_bn(seq, 413)
>>> np.all(ret1-ret2 == 0)
True
```

I moved the linetools dependency from pip to conda since the 0.3dev2231 is on [conda-forge](https://anaconda.org/conda-forge/linetools/files).

I think the attribution in the `fast_running_median` docstring may need to be updated, but I would like to hear others' thoughts on that.

### CODE REVIEW

**Review by jhennawi — CHANGES_REQUESTED** [review]
Hi @finagle29, 

Thanks for figuring this out. This definitely looks preferably to the python code we have since it is compiled C code, so I'm not surprised it is faster. I'll defer to others like @profxj about the added dependency on bottleneck, although it seems like a useful package. 

My main concern is how the boundaries are treated. It is very important that this function treat the boundaries via reflection and not via some other approach. This is clearly explained in both the SciPy and current fast_running_median, however, it is very unclear in the bottleneck docs how boundaries are treated. 

If the bottleneck version does not have this functionality, we can easily add it ourselves but some care is required there.

**Comment by finagle29** [comment]
The boundaries are treated as they were before, by padding the input array to produce the desired reflected boundaries. The `bottleneck.move_median` function is easily configured to produce NaNs at the boundaries and to ignore data off the edge (neither of which correspond to scipy boundary conditions) but the boundary-handling code that you wrote and this PR keeps ensures that the boundaries are handled correctly.

File: pypeit/utils.py
Comment (kbwestfall) [inline pypeit/utils.py]: The use of `as` is a pet-peeve of mine.  If this is an "industry standard", then I'll get over it.  Otherwise, I'd prefer we just do `import bottleneck` or `from bottleneck import move_median`.

**Review by kbwestfall — APPROVED** [review]
Looks good!  I have one minor comment (that likely no one but me cares about), but I think this is a good substitution, as long as @jhennawi is satisfied with your response about the boundaries.  I think adding bottleneck as a dependency is fine, and note that it's an optional dependency for astropy.  We could follow suite, but I'm fine with having the package be an explicit dependency.  I would update the docstring for `fast_running_median` to reflect the change in the method.  I'm approving now, but we should run the tests and the dev-suite just to make sure this doesn't have any unintended consequences.

File: pypeit/utils.py
Comment (jhennawi) [reply pypeit/utils.py]: I agree @kbwestfall, except for standard abbreviations like np for numpy and plt for pyplot.

Comment (finagle29) [reply pypeit/utils.py]: bottleneck's github README uses `as`, which is why I used `as`, but Astropy doesn't so I'm happy to change it to just `import bottleneck`

**Comment by jhennawi** [comment]
Okay, I see now how the boundary is dealt with. Can you please update the docs. I agree we should re-run tests.

**Comment by profxj** [comment]
@finagle29 -- Let me know if you wish for me to start the tests.
Happy to use my machine for this.

**Comment by finagle29** [comment]
@profxj yes please!

**Comment by profxj** [comment]
Unit tests pass

================ 258 passed, 391 warnings in 1768.65s (0:29:28) ================

**Comment by profxj** [comment]
Two DevSuite failures:

--- PYPEIT DEVELOPMENT SUITE FAILED 2/100 TESTS (Masters ignored) ---

2 passed/ 1 failed/ 0 skipped FAILED  keck_deimos/830G_M_8500 pypeit (without masters)
----
keck_deimos/830G_M_8500 pypeit (without masters) Result: --- FAILED

Logfile:    /scratch/PypeIt-development-suite/REDUX_OUT/keck_deimos/830G_M_8500/keck_deimos_830g_m_8500.test.log
Process Id: 263339
Start time: Tue Mar 30 16:15:11 2021
End time:   Tue Mar 30 17:14:09 2021
Duration:   0:58:58.043532
Command:    run_pypeit /scratch/PypeIt-development-suite/REDUX_OUT/keck_deimos/830G_M_8500/keck_deimos_830g_m_8500.pypeit -o

Error Messages:

End of Log:
  File "/data/Projects/Python/PypeIt/pypeit/pypmsgs.py", line 184, in error
    raise PypeItError(msg)
pypeit.pypmsgs.PypeItError: No valid data points in bspline_profile!.

The other is the 'No matching objects for OBJ0205-DET01' failure.

@finagle29 -- Can you try the DEIMOS test on your machine?

File: pypeit/core/skysub.py
Comment (jhennawi) [inline pypeit/core/skysub.py]: Can you please make the following changes here: 1) rename inmask_in to gpm everywhere in global_skysub to be more clear. This is a change we are trying to propagate everywhere. 2) print out a more informative warning message like "Input gpm pixel mask is eveyrwhere False for this slit.  There is likely a problem with this slit."

**Review by jhennawi — APPROVED** [review]
Please make the changes above and then merge.

File: pypeit/core/skysub.py
Comment (profxj) [reply pypeit/core/skysub.py]: done.  am re-running DEIMOS tests just to be sure..

**Comment by profxj** [comment]
================ 258 passed, 391 warnings in 1814.30s (0:30:14) ================

**Comment by profxj** [comment]
DEIMOS is passing.  

10 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/1200G_M_5500 pypeit (without masters)

merging..

**Pull request merged**
